### DISCUSSION TITLE
fabric.Textbox padding property issues

### DISCUSSION THREAD

**Discussion by rwmorton** [fabricjs/fabric.js#8521] [General]
Hi there,

I'm having a little trouble using the padding property on the Textbox. By default it works fine, if I scale in any direction the padding stays correctly offset. I am using Fabric 5.2.4.

![Padding property works fine](https://user-images.githubusercontent.com/95928533/208446779-2f5f764b-1807-48e9-8a93-5e46ff725627.png)
![Padding property works fine even when scaling](https://user-images.githubusercontent.com/95928533/208446808-0245fc3d-65cf-4537-baeb-184eade3c3ae.png)

What I want to do is to fill in the padded area with color. The issue is the control doesn't seem to work properly. It works fine ONLY if scaling in the horizontal direction. Not when scaling in Y or using the corner control. It also doesn't work when zooming the canvas.

![Rectangle correctly fills the padding space](https://user-images.githubusercontent.com/95928533/208447193-1def6982-329b-4ea4-8cbc-2167bdf24b66.png)
![Also works fine when scaling horizontally](https://user-images.githubusercontent.com/95928533/208447243-e8c15162-652f-4182-9e19-2d94b4307ce2.png)
![But the controls go whack when scaling in any other direction or zooming the canavs](https://user-images.githubusercontent.com/95928533/208447262-8c87be95-f009-4584-a756-7773c954e7e1.png)

```
import { fabric } from 'fabric'

export const StickyNote = fabric.util.createClass(fabric.Textbox,{
  type: 'stickyNote',

  toObject: function() {
    return fabric.util.object.extend(this.callSuper('toObject'),{
      backgroundColor: this.get('backgroundColor'),
      padding: this.get('padding')
    })
  },

  _renderBackground: function(ctx) {
    if(!this.backgroundColor) {
      return
    }

    const dim = this._getTransformedDimensions()
    ctx.fillStyle = this.backgroundColor

    ctx.fillRect(
      -dim.x / 2 - this.padding,
      -dim.y / 2 - this.padding,
      dim.x + this.padding * 2,
      dim.y + this.padding * 2
    )
    this.setCoords()
    // this._removeShadow(ctx)
  },
})

StickyNote.fromObject = function(object,callback) {
    return fabric.Object._fromObject('StickyNote',object,'text')
}

const note = new StickyNote('This is a Textbox',{
    textAlign: 'center',
    top: 100,
    left: 100,
    width: 200,
    fontSize: 30,
    fontWeight: 'bold',
    fill: 'black',
    backgroundColor: 'yellow',
    padding: 20
  })

  canvas.add(note)
```

**Comment by ShaMan123** [comment]
`_renderBackground`, as all object rendering methods (not controls), runs on a context transformed to the object's coordinate plane.
Meaning, the context is already scaled, rotated, skewed etc. and is positioned so that (0, 0) is at the object's center.
Applying scale in this context is wrong.
https://jsfiddle.net/8unw6mLe/

**Reply by rwmorton** [reply]
Thank you very mcuh @ShaMan123 !

**Comment by rwmorton** [comment]
Hi there, how do I correctly update the bounding box coordinates to match that of the context.fillRect area?

This works fine if the padding property === 0, but not if > 0.

Here it is with no scale - that is getTotalObjectScaling === (1, 1)

![no-scale](https://user-images.githubusercontent.com/95928533/212853095-ca4864ee-0b76-496e-8d8c-120b10f74432.png)

getTotalObjectScaling() ~ (3, 3)

![scaled-approx-300-percent](https://user-images.githubusercontent.com/95928533/212853129-57ab9a86-0636-4e8f-80e2-0d0057aa1efe.png)

getTotalObjectScaling() ~ (0.3, 0.3)

![scaled-approx-30-percent](https://user-images.githubusercontent.com/95928533/212853156-68ac530b-6bbf-456e-8202-019d4ae9bc64.png)

How do I update the controls correctly so that they fit the filled rectangle?

**Comment by rwmorton** [comment]
Bump

**Comment by rwmorton** [comment]
How do I also factor in the viewport transform (scale)?

**Comment by rwmorton** [comment]
Sorry I just can't seem to get this working!

![image](https://user-images.githubusercontent.com/95928533/216048781-785fcf5e-cde9-4ac3-8154-d8bedb03a974.png)

Red box is a rectangle that should be scaling correctly with the viewport. Scaling down from 100% seems to work fine, but scaling upward of that and things break. Also the controls box doesn't fit the note (yellow box).

[Peek 2023-02-01 14-53.webm](https://user-images.githubusercontent.com/95928533/216048568-aa11408e-e057-4e6b-b7f9-a8ebebf4919b.webm)

**Comment by ShaMan123** [comment]
Well you are subclassing so you are hitting fabric core, diving into the complex logic of transformations.
Maybe you can consider using Group in v6? Wrap the textbox with Group, handle layout in Group and you are done...
Otherwise you need to adjust controls (though I thought padding affects controls, maybe there is something already doing that, look into the control source code)

**Reply by ShaMan123** [reply]
And maybe play around with a simple class like rect or circle and get it working there before tackling nasty complex Textbox

**Comment by rwmorton** [comment]
Thanks for the feedback. Sorry for the late response got caught up with another project. So the problem is the viewport transform somehow, as you can see I've manually offset the controls and they match (with a little error which might be sync issues).

So basically I need to get the controls to include the viewport transform somehow. Any ideas?

CONTROLS MATCH RECTANGLE

[Peek 2023-02-06 07-20.webm](https://user-images.githubusercontent.com/95928533/216890374-24679634-beeb-4bc9-9cbc-fe9e23f03a30.webm)

BROKEN CONTROLS

[Peek 2023-02-06 07-22.webm](https://user-images.githubusercontent.com/95928533/216890392-f0b4e2f9-89ac-4af5-8dbf-846bcae7d315.webm)

MANUALLY ADJUSTING CONTROLS

![image](https://user-images.githubusercontent.com/95928533/216890332-115cfdd3-43d8-475b-bd0a-273abf93ab76.png)

**Reply by ShaMan123** [reply]
better to use `Control#positionHandler` and call setCoords once vpt changes and you are done
